### DISCUSSION TITLE
pytest_plugins in root conftest.py not working with tox

### DISCUSSION THREAD

**Discussion by fschulze** [pytest-dev/pytest#10324] [Q&A] [answered]
Hi!

In devpi-server are several fixtures which are reused in plugin tests. That isn't optimal, see https://github.com/devpi/devpi/issues/780. Some other fixtures are currently reused by importing them from ``test_devpi_server.conftest``, some other with ``pytest_devpi_server`` which is a plugin with an entry point. The direct importing also seems to be bad for various reasons.

After trying for some time I managed to get it working by using ``pytest_plugins = ["pytest_devpi_server", "test_devpi_server.conftest"]`` in the root ``conftest.py`` of devpi-web.

Problem is, it only works when using ``pytest`` directly, not when using ``tox``. I tried to debug it and am stuck. The modules are reached, which is confirmed by both ``tox -e py37 -- --trace-config`` and sprinkling ``pdb.set_trace()`` in them. All relevant modules ``pytest_devpi_server``, ``test_devpi_server.conftest``, ``web/conftest.py`` and ``web/tests/conftest.py`` are reached and included in the trace. Still the fixtures from ``test_devpi_server.conftest`` are not found when running tox, but work fine when using pytest directly.

**Comment by portyu9** [comment, marked as answer]
Hi @fschulze 

I would avoid using another package's `conftest.py` itself as the reusable plugin.

Instead of:

```python
pytest_plugins = [
    "pytest_devpi_server",
    "test_devpi_server.conftest",
]
```

move the shared fixtures into a normal importable module, for example:

```text
test_devpi_server/fixtures.py
```

and load that:

```python
pytest_plugins = [
    "pytest_devpi_server",
    "test_devpi_server.fixtures",
]
```

Or, better still, expose those fixtures from the actual `pytest_devpi_server` plugin if they belong to that plugin.

`conftest.py` has special pytest discovery and directory-scoping semantics, so treating one as a reusable cross-package plugin makes behavior unnecessarily dependent on how the source tree is laid out and imported. A normal plugin module avoids that ambiguity and behaves consistently from pytest and tox.

Since your `--trace-config` output says the modules are being loaded under tox, I would not conclude that tox simply failed to load `pytest_plugins`. The next useful diagnostic would be:

```text
tox -e py37 -- pytest --fixtures -v
```

and compare whether the expected fixture names are actually registered and where pytest says they come from.

The robust fix is nevertheless to move reusable fixtures out of `conftest.py` and into a normal plugin module.

Please mark this answer as accepted if it helped, thank you.

**Reply by fschulze** [reply]
Thanks for taking the time to answer this. I already fixed it as you described about a year later. I hope this helps others as well.
